Fixes to some issues reported by static analyzer in tracking related code

makortel opened the pull request:
This PR attempts to fix some of the issues reported by static analyzer in tracking (and related) code.

Tested in CMSSW_7_6_X_2015-06-30-2300, no changes expected in standard workflows.

@rovere @VinInn

makortel committed:
Replace C-style casts with static_cast, preserve const

makortel committed:
Remove bunch of dead assignments

makortel committed:
Fix finding chi2min in CRackTrajectoryBuilder

Found with static analyzer

makortel committed:
Check pointers for null before their use

makortel committed:
Use unique_ptr in RectangularEtaPhiTrackingRegion to silence static analyzer

makortel committed:
Use edm::isNotFinite in DeterministicAnnealing and GeometricAnnealing

Suggested by static analyzer

cmsbuild commented:
A new Pull Request was created by @makortel (Matti Kortelainen) for CMSSW_7_6_X.

Fixes to some issues reported by static analyzer in tracking related code

It involves the following packages:

DQM/SiStripMonitorTrack
RecoTracker/ConversionSeedGenerators
RecoTracker/FinalTrackSelectors
RecoTracker/SingleTrackPattern
RecoTracker/TkHitPairs
RecoTracker/TkSeedGenerator
RecoTracker/TkTrackingRegions
RecoTracker/TrackProducer
RecoVertex/BeamSpotProducer
RecoVertex/VertexTools
SimG4CMS/Tracker
SimGeneral/TrackingAnalysis
SimTracker/TrackerHitAssociation
TrackPropagation/SteppingHelixPropagator
TrackingTools/TrackAssociator
Validation/RecoTrack
Validation/RecoVertex


@civanch, @diguida, @cvuosalo, @mdhildreth, @cmsbuild, @cerminar, @deguio, @slava77, @mmusich, @danduggan can you please review it and eventually sign? Thanks.
@ghellwig, @battibass, @threus, @abbiendi, @GiacomoSguazzoni, @jhgoh, @VinInn, @bellan, @istaslis, @mschrode, @rovere, @wmtford, @gpetruc, @cerati, @trocino, @dgulhan, @venturia this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

diguida commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

deguio commented:
+1

mmusich commented:
+1 
minor changes in BSFitter code and TrackerHitFilter affecting AlCa

slava77 commented:
@cmsbuild please test
I don't see this test in jenkins lineup

VinInn commented on RecoVertex/VertexTools/src/DeterministicAnnealing.cc:
```diff
@@ -36,7 +37,7 @@ double DeterministicAnnealing::weight ( double chi2 ) const
   */
   // return mphi / ( mphi + phi ( theChi2cut ) );
   long double newtmp = mphi / ( mphi + phi ( theChi2cut ) );
```
long double???
this is a joke  (next round we should use "double")

slava77 commented:
I suppose "some " in the subject of this PR is the key ;)
there are more things reported even in the files touched here
https://slava77sk.web.cern.ch/slava77sk/reco/llvm/10066/
I suggest further changes are done in the next PR

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/10066	5f95bc2
- since jenkins is stuck, tested in CMSSW_7_6_X_2015-07-04-2300 /test area sign570/
- the static analyzer warnings were cleaned up considerably https://slava77sk.web.cern.ch/slava77sk/reco/llvm/10066/
- short matrix tests pass and comparisons with baseline show no difference

cmsbuild commented:
-1 
Tested at: 5f95bc2029a2b2c96df08d3e3ff6904d571abe8c
 I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 8 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_6_X_2015-07-08-2300/src/RecoTracker/TkHitPairs/src/InnerDeltaPhi.cc:22:17: warning: unused function 'checked_asin' [-Wunused-function]
  inline double checked_asin(double x, const char *expr, const char *file, int line) {
                ^
1 warning generated.
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_6_X_2015-07-08-2300/src/RecoTracker/TrackProducer/src/TrackProducerAlgorithm.cc 
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_6_X_2015-07-08-2300/src/RecoTracker/TkHitPairs/src/HitPairGeneratorFromLayerPair.cc:139:15: error: variable-sized object may not be initialized
      bool ok[e-b]{};
              ^~~
1 error generated.
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_6_X_2015-07-08-2300/src/RecoTracker/TrackProducer/src/TrajectoryToResiduals.cc 
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_6_X_2015-07-08-2300/src/RecoTracker/TkTrackingRegions/src/SealModule.cc:5:
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10066/6084/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/b249178a4b1948d1d408753cae2f9fb1f8e91238
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10066/6084/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10066/6084/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10066/6084/summary.html

makortel commented:
Of course clang wasn't happy about that... I can take a look on it next week.

VinInn commented:
I have a final solution
https://github.com/VinInn/cmssw/blob/6127ed0ab28919992e4c072bde3cebd70b1319fc/CommonTools/Utils/interface/DynArray.h
Maybe I will make a PR with only that...

makortel commented:
Thanks, that should do it (I was also thinking that I saw something related in some your PRs). Please go ahead if you want, I can then adjust this one as needed.

makortel committed:
Use enum class for hit match status in SiStripTrackingRecHitsValid

Hopefully this will convince static analyzer that matchedmonorstereo
does not change within the rechitanalysis_matched(). Anyway, we get
rid of some (repeated) string comparisons.

makortel commented:
I removed the commit causing problems for clang (\*), switched string comparisons to (const) strongly-typed enum in SiStripTrackingRecHitsValid (**), and rebased on top of CMSSW_7_6_X_2015-07-13-1100.

(\*) while waiting for Vincenzo's DynArray
(**) in the hope that now static analyzer would not assume that the content of `matchedmonorstereo` could change between the comparisons (I say "hope" because for me `scram b checker` crashes for most of the files, including SiStripTrackingRecHitsValid.cc)

cmsbuild commented:
Pull request #10066 was updated. @civanch, @diguida, @cvuosalo, @mdhildreth, @cmsbuild, @cerminar, @deguio, @slava77, @mmusich, @danduggan can you please check and sign again.

deguio commented:
+1

cmsbuild commented:
-1 
Tested at: 27386989bde1f95cdb0382c9ee43531f96a3c96f
 I found an error when building: 
 
 <pre>>> Building binary trackAlgoPriorityOrder_t
>> Building binary FastCircle_t
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsMath
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsCommon
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsStdDictionaries
collect2: error: ld returned 1 exit status
>> Deleted: tmp/slc6_amd64_gcc491/src/RecoTracker/TkSeedGenerator/test/FastCircle_t/FastCircle_t
gmake: *** [tmp/slc6_amd64_gcc491/src/RecoTracker/TkSeedGenerator/test/FastCircle_t/FastCircle_t] Error 1
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsTrackReco
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsTrackCandidate
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsTrackingRecHit
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10066/6273/summary.html

cmsbuild commented:
-1 
Tested at: 27386989bde1f95cdb0382c9ee43531f96a3c96f
 I found an error when building: 
 
 <pre>/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsDetId
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsMath
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsSiStripCluster
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsCommon
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsStdDictionaries
collect2: error: ld returned 1 exit status
>> Deleted: tmp/slc6_amd64_gcc491/src/RecoTracker/FinalTrackSelectors/test/trackAlgoPriorityOrder_t/trackAlgoPriorityOrder_t
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsMath
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsCommon
/afs/cern.ch/cms/sw/ReleaseCandidates/volC/slc6_amd64_gcc491/external/gcc/4.9.1-cms/bin/../lib/gcc/x86_64-redhat-linux-gnu/4.9.1/../../../../x86_64-redhat-linux-gnu/bin/ld: cannot find -lDataFormatsStdDictionaries
collect2: error: ld returned 1 exit status
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10066/6276/summary.html

slava77 commented:
@cmsbuild please test
IBs are OK now

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 27386989bde1f95cdb0382c9ee43531f96a3c96f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10066/6347/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10066/6347/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/10066	2738698
- compared to the previous "+1" for 5f95bc2, changes are only to back-out a change (revert to baseline) in HitPairGeneratorFromLayerPair.cc [related to reco code] and in SiStripTrackingRecHitsValid.h [only in validation]
- jenkins tests passed and comparisons with baseline show no (relevant) differrence
